include/bout/petsc_interface.hxx

```diff
@@ -42,64 +42,223 @@
 #include <bout/traits.hxx>
 #include <bout_types.hxx>
 #include <boutcomm.hxx>
+#include <bout/operatorstencil.hxx>
 
 #ifdef BOUT_HAS_PETSC
+template <class T>
 class GlobalIndexer;
-using IndexerPtr = std::shared_ptr<GlobalIndexer>;
+
+template <class T>
+using IndexerPtr = std::shared_ptr<GlobalIndexer<T>>;
+
 using InterpolationWeights = std::vector<ParallelTransform::PositionsAndWeights>;
 
 /*!
  * A singleton which accepts index objects produced by iterating over
  * fields and returns a global index. This index can be used when
  * constructing PETSc arrays. Guard regions used for communication
  * between processes will have the indices of the part of the interior
- * region they are mirroring. Boundaries will have unique indices, but
- * are only included to a depth of 1.
+ * region they are mirroring. Boundaries required by the stencil will
+ * have unique indices. If no stencil is provided then only the
+ * interior region will be assigned global indices. By default, the
+ * indexer is fully initialised so that guard cells are communicated
+ * (ensuring they hold the appropriate global indices). However, by
+ * passing `autoInitialise = false` this behaviour can be turned off
+ * and the user can then manually call the `initialise()` method
+ * later. This can be useful for mocking/faking the class when
+ * testing.
  */
+template<class T>
 class GlobalIndexer {
 public:
-  /// If \p localmesh is the same as the global one, return a pointer
-  /// to the global instance. Otherwise create a new one.
-  static IndexerPtr getInstance(Mesh* localmesh);
+  static_assert(bout::utils::is_Field<T>::value, "GlobalIndexer only works with Fields");
+  using ind_type = typename T::ind_type;
+
+  GlobalIndexer() = default;
+  
+  GlobalIndexer(Mesh* localmesh,
+                OperatorStencil<ind_type> stencil = OperatorStencil<ind_type>(),
+		bool autoInitialise = true)
+    : fieldmesh(localmesh), indices(-1., localmesh), stencils(std::move(stencil)) {
+
+    Region<ind_type> allCandidate, bndryCandidate;
+    if (stencils.getNumParts() > 0) {
+      std::set<ind_type> allIndices(getRegionNobndry().getIndices().begin(),
+				    getRegionNobndry().getIndices().end()),
+	newIndices;
+      BOUT_FOR_SERIAL(i, getRegionNobndry()) {
+	for (const IndexOffset<ind_type> j : stencils.getStencilPart(i)) {
+	  insertIndex(i+j, allIndices, newIndices);
+	}
+      }
+      std::set<ind_type> candidateIndices = newIndices;
+      while (candidateIndices.size() > 0) {
+	newIndices.clear();
+        for (const ind_type i : candidateIndices) {
+	  insertIndex(i, allIndices, newIndices);
+        }
+	candidateIndices = newIndices;
+      }
+      std::vector<ind_type> allIndicesVec(allIndices.begin(), allIndices.end());
+      allCandidate = Region<ind_type>(allIndicesVec);
+    } else {
+      regionAll = getRegionNobndry();
```

Comment:
`regionAll` is unconditionally assigned to below (L120); is this value used?

Reply:
Thanks for catching that. It was a remnant of an older (incorrect) algorithm I was using.